### PR TITLE
media-libs/mesa: add nvk support to mesa-9999

### PR DESCRIPTION
**Pull request by MrDuartePT**
This commit finnaly add mesa support for nvk, this way when mesa-24.1.0 or when using 9999 the compile will not fail with nouveau flag enable.

This commit also fix rustc not capable of compiling programs on multilib profile, more info how this was achieved here: https://gitlab.freedesktop.org/mesa/mesa/-/issues/10490

### CODE REVIEW

**Comment by mattst88** [comment]
cc: @gentoo/rust 

Is this the appropriate way to handle rust dependencies?

**Comment by MrDuartePT** [comment]
@mattst88 I know rust.eclass is commonly use to handle rust depedencies, but I think it not possible to use in this case. Since mesa don't provide a Cargo.toml

**Comment by MrDuartePT** [comment]
It meson-1.4.0 introduce a regression when compiling NVK, more info here: https://gitlab.freedesktop.org/mesa/mesa/-/issues/10855

File: media-libs/mesa/mesa-9999.ebuild [hunk 1]
```diff
@@ -127,6 +143,7 @@ DEPEND="${RDEPEND}
 		x11-base/xorg-proto
 	)
 "
+#Meson regression nouveau/NVK (upstream bug #10855)
```
Comment (Jamim) [inline media-libs/mesa/mesa-9999.ebuild:146]: _Hello @MrDuartePT,_

I believe it's worth providing direct links for better transparency.
```suggestion
# meson-1.4.0 contains a regression, so it fails to compile nouveau/NVK
# see https://gitlab.freedesktop.org/mesa/mesa/-/issues/10855
```

File: media-libs/mesa/mesa-9999.ebuild [hunk 2]
```diff
@@ -237,6 +267,21 @@ src_prepare() {
 	default
 	sed -i -e "/^PLATFORM_SYMBOLS/a '__gentoo_check_ldflags__'," \
 		bin/symbols-check.py || die # bug #830728
+
+	if use video_cards_nouveau; then
+		# NVK Subproject Handeling
```
Comment (Jamim) [inline media-libs/mesa/mesa-9999.ebuild:272]:
```suggestion
		# NVK Subproject Handling
```

File: media-libs/mesa/mesa-9999.ebuild [hunk 3]
```diff
@@ -237,6 +267,21 @@ src_prepare() {
...
+		pushd "${WORKDIR}" || die
+			for subpkg in proc-macro2-${proc_macro2_PV} syn-${syn_PV} quote-${quote_PV} unicode-ident-${unicode_ident_PV}; do
+				# copy subproject folder
+				subpkg_folder=mesa-${PV}/subprojects
+				cp -r ${subpkg} ${subpkg_folder} || die
+				# copy meson.build
+				cp ${subpkg_folder}/packagefiles/${subpkg%-*}/meson.build ${subpkg_folder}/${subpkg} || die
+				#ovewrite subpkg version when needed
```
[the remark this answers is not in the document]
Comment (MrDuartePT) [reply media-libs/mesa/mesa-9999.ebuild:280]: Ok gonna do the changes thanks

**Comment by MrDuartePT** [comment]
@Jamim and @mattst88 it something else need to done for this to get accepted.
Or gentoo team is waiting for mesa 24.1?

**Comment by TyroneWatermelon420** [comment]
@MrDuartePT i just tried compiling it but its throwing "Run-time dependency paste found: NO (tried pkgconfig and cmake)" at me, maybe you should include it into NAK_URI

**Comment by MrDuartePT** [comment]
> @MrDuartePT i just tried compiling it but its throwing "Run-time dependency paste found: NO (tried pkgconfig and cmake)" at me, maybe you should include it into NAK_URI

Done, you can try it now.

**Comment by MrDuartePT** [comment]
@mattst88 since 24.1.0_rc1 was released and this ebuild is stable enough (I being used for a long time) we can merge this after I rebased to 24.1.0_rc1 as well?

**Comment by MrDuartePT** [comment]
The checksum overwrite is fix

**Comment by mattst88** [comment]
Please squash the non-meson-1.4.0 changes from [media-libs/mesa: fix regression on => dev-build/meson-1.4.0 (NVK only)](https://github.com/gentoo/gentoo/pull/35658/commits/1d7033b9a232798d9f8c110090cb8758afb7a8de) into the initial patch.

Please squash [media-libs/mesa: add paste submodule to NAK_URI](https://github.com/gentoo/gentoo/pull/35658/commits/1f445b406fd01eedd30cdfb9d9eeb8afdc8756a3) into the initial patch.

(I can do these two if you don't know how)

Let's drop [media-libs/mesa: drop 24.1.0_rc1; add 24.1.0_rc1-r1](https://github.com/gentoo/gentoo/pull/35658/commits/8ab5535535d5fe97225561de81fa40ccd60c5d6f) -- I'll copy from -9999 when I do the version bump to 24.1.0_rc2.

**Comment by mattst88** [comment]
I feel like at least initially I want us to have a `VIDEO_CARDS=nvk` setting that we can `package.use.mask` separately (because we can't `package.use.mask` `VIDEO_CARDS=nouveau` **and** `USE=vulkan` without masking them individually). I suspect we're going to have some bugs enabling this new code.

**Comment by mattst88** [comment]
```
>>> Preparing source in /var/tmp/portage/media-libs/mesa-24.1.0_rc1-r1/work/mesa-24.1.0-rc1 ...
/var/tmp/portage/media-libs/mesa-24.1.0_rc1-r1/work /var/tmp/portage/media-libs/mesa-24.1.0_rc1-r1/work/mesa-24.1.0-rc1
cp: cannot create directory 'mesa-24.1.0_rc1/subprojects': No such file or directory
```

Fails to src_prepare for me. Looks like we need a `mkdir -p` or something.

**Comment by MrDuartePT** [comment]
> Please squash the non-meson-1.4.0 changes from [media-libs/mesa: fix regression on => dev-build/meson-1.4.0 (NVK only)](https://github.com/gentoo/gentoo/pull/35658/commits/1d7033b9a232798d9f8c110090cb8758afb7a8de) into the initial patch.
> 
> Please squash [media-libs/mesa: add paste submodule to NAK_URI](https://github.com/gentoo/gentoo/pull/35658/commits/1f445b406fd01eedd30cdfb9d9eeb8afdc8756a3) into the initial patch.
> 
> (I can do these two if you don't know how)
> 
> Let's drop [media-libs/mesa: drop 24.1.0_rc1; add 24.1.0_rc1-r1](https://github.com/gentoo/gentoo/pull/35658/commits/8ab5535535d5fe97225561de81fa40ccd60c5d6f) -- I'll copy from -9999 when I do the version bump to 24.1.0_rc2.

Ok I gonna do all the changes

**Comment by MrDuartePT** [comment]
> I feel like at least initially I want us to have a `VIDEO_CARDS=nvk` setting that we can `package.use.mask` separately (because we can't `package.use.mask` `VIDEO_CARDS=nouveau` **and** `USE=vulkan` without masking them individually). I suspect we're going to have some bugs enabling this new code.

I can see this causing problems if a user have a card that don't support NVK (I dont have older hardware to test), so it probably better to that way (well I probably will edit the wiki when have more time to reflect that flag).

For my use with both Nouveau and Nvidia driver seems to work fine. But, I have setup my kernel command line (using refind), to mask the driver is not in use:

NVK drivers: ```"Normal Boot Nouveau" "... modprobe.blacklist=nvidia nouveau.modeset=1 ..."```

Nvidia drivers: ```"Normal Boot Nvidia" "... modprobe.blacklist=nouveau nvidia_drm.modeset=1 nvidia_drm.fbdev=1 nvidia_modeset.hdmi_deepcolor=1 nvidia.NVreg_PreserveVideoMemoryAllocations=0 nvidia.NVreg_DynamicPowerManagement=0x03 ..."```

**Comment by MrDuartePT** [comment]
> ```
> >>> Preparing source in /var/tmp/portage/media-libs/mesa-24.1.0_rc1-r1/work/mesa-24.1.0-rc1 ...
> /var/tmp/portage/media-libs/mesa-24.1.0_rc1-r1/work /var/tmp/portage/media-libs/mesa-24.1.0_rc1-r1/work/mesa-24.1.0-rc1
> cp: cannot create directory 'mesa-24.1.0_rc1/subprojects': No such file or directory
> ```
> 
> Fails to src_prepare for me. Looks like we need a `mkdir -p` or something.

I think that is happening because of the way the subprojects are being copy but let me check the tarball created by mesa, (if the folder dosen't exist this can create problems because the wrap files are required for checking if the dependencies are present and in the right location and the meson files to properly build them).

**Comment by MrDuartePT** [comment]
`VIDEO_CARDS=nvk` was added, and also make sure the driver is only build if `VIDEO_CARDS=nouveau` is also set.

**Comment by mattst88** [comment]
I pushed the first patch and intended to push the second, but I can't get the multilib build to work. I'm pretty sure it was working for me last week.

It doesn't look like this could be the result of any changes to the PR, but I notice that I updated from rust-1.75.0-r1 to rust-1.77.1 since I last tested.

Does the multilib stuff need some change for rust-1.77?

Also, I see that https://gitlab.freedesktop.org/mesa/mesa/-/issues/10490 is closed. Is this because we determined that there was nothing to fix in upstream Mesa?

**Comment by MrDuartePT** [comment]
> I pushed the first patch and intended to push the second, but I can't get the multilib build to work. I'm pretty sure it was working for me last week.
> 
> 
> 
> It doesn't look like this could be the result of any changes to the PR, but I notice that I updated from rust-1.75.0-r1 to rust-1.77.1 since I last tested.
> 
> 
> 
> Does the multilib stuff need some change for rust-1.77?
> 

Being fare I didn't test it yet, I don't really sure if my system have rust up to date. I gonna try it when I get on the system.

> 
> Also, I see that https://gitlab.freedesktop.org/mesa/mesa/-/issues/10490 is closed. Is this because we determined that there was nothing to fix in upstream Mesa?

Yes nothing need to be changed in Mesa upstream the bug was more of a gentoo specify thing. I only open upstream because I was having problems compile from source directly (because of the rust need a seperate cross file)

**Comment by MrDuartePT** [comment]
@mattst88 found the problem new dependencie is needed: dev-util/cbindgen:
```
Compiler for C supports arguments -Wno-gnu-variable-sized-type-not-at-end: YES
../mesa-9999/src/nouveau/nil/meson.build:7: WARNING: add_languages is missing native:, assuming languages are wanted for both host and build.
Program cbindgen found: NO

../mesa-9999/src/nouveau/nil/meson.build:15:16: ERROR: Program 'cbindgen' not found or not executable
```
This depedency was introduce with this commit: https://gitlab.freedesktop.org/mesa/mesa/-/commit/e297fb1bd59d08a5df8f524ab47c7bd84b39a551

**Comment by MrDuartePT** [comment]
@mattst88 Well for some weird reason dev-util/cbindgen disapear from my system, and that why I also getting a error now.
After emerging it I can confirm at least on my side that mesa-9999 compile without any problems with `rust-bin-1.77.1`.

**Comment by MrDuartePT** [comment]
I also noticed you bumped dev-util/spirv-llvm-translator to 18.1 two weak ago, want me to test if the mesa builds with it?

**Comment by mattst88** [comment]
Yeah, that would be great.

**Comment by MrDuartePT** [comment]
> Yeah, that would be great.

After changing some gst-plugins packages useflag (because of wine dependencies and now mesa-9999 only relay on opengl useflag), might make a PR to change that (the problem is most of GStreamer stuff ebuilds are stable).
I can confirm `mesa-9999` compile fine with `dev-util/spirv-llvm-translator-18.1.0`:
```bash
 * dev-util/spirv-llvm-translator

     Wed May  8 00:15:55 2024 >>> dev-util/spirv-llvm-translator-18.1.0
       merge time: 50 seconds.

 * media-libs/mesa
 
     Wed May  8 00:26:41 2024 >>> media-libs/mesa-9999
       merge time: 10 minutes and 46 seconds.
```
Matt you didn't confirm if mesa ebuild still work on your side, since I not get any error on my system.

**Comment by mattst88** [comment]
It still fails for me:

```
FAILED: src/nouveau/vulkan/libvulkan_nouveau.so 
x86_64-pc-linux-gnu-g++ -m32 -mfpmath=sse  -o src/nouveau/vulkan/libvulkan_nouveau.so  -flto -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,libvulkan_nouveau.so -Wl,--whole-archive -Wl,--start-group src/nouveau/vulkan/libnvk.a -Wl,--no-whole-archive -O2 -march=tigerlake -mno-avx512f -fcf-protection=full -pipe -fno-omit-frame-pointer -Wl,-O1 -Wl,--as-needed -Wl,-z,pack-relative-relocs src/nouveau/codegen/libnouveau_codegen.a src/util/libmesa_util.a src/util/libmesa_util_sse41.a src/util/blake3/libblake3.a src/c11/impl/libmesa_util_c11.a src/nouveau/compiler/libnak.a src/nouveau/compiler/libnak_rs.a src/nouveau/compiler/libbitview.rlib src/nouveau/compiler/libnak_bindings.rlib src/nouveau/compiler/libnak_ir_proc.so subprojects/paste-1.0.14/libpaste.so src/nouveau/headers/libnvidia_headers.rlib src/nouveau/headers/libnvidia_headers_c.a src/nouveau/nil/liblibnil.a src/nouveau/nil/liblibnil_format_table.a src/nouveau/nil/libnil_rs_bindings.rlib src/compiler/nir/libnir.a src/compiler/libcompiler.a src/nouveau/mme/libnouveau_mme.a src/nouveau/winsys/libnouveau_ws.a src/vulkan/util/libvulkan_util.a src/compiler/spirv/libvtn.a src/loader/libloader_wayland_helper.a src/util/libxmlconfig.a -Wl,--build-id=sha1 /usr/lib/libdrm.so /usr/lib/libz.so -pthread -lm /usr/lib/libzstd.so /usr/lib/libxcb.so /usr/lib/libX11-xcb.so /usr/lib/libX11.so /usr/lib/libxcb-dri3.so /usr/lib/libxcb-present.so /usr/lib/libxcb-xfixes.so /usr/lib/libxcb-sync.so /usr/lib/libxcb-randr.so /usr/lib/libxcb-shm.so /usr/lib/libxshmfence.so /usr/lib/libxcb-keysyms.so /usr/lib/libwayland-client.so /usr/lib/libXrandr.so /usr/lib/libudev.so -lrt -lpthread -ldl -lutil -lutil /usr/lib/libexpat.so -Wl,--end-group
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: src/nouveau/compiler/libnak_ir_proc.so: error adding symbols: file in wrong format collect2: error: ld returned 1 exit status
```

I'll try a few things and report back.

**Comment by mattst88** [comment]
Rebuilding cbindgen didn't help. Going to go back to rust-1.75 and check.

**Comment by MrDuartePT** [comment]
> ```
> /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86 # file src/nouveau/compiler/libnak_ir_proc.so
> src/nouveau/compiler/libnak_ir_proc.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, not stripped
> ```
> 
> So `libnak_ir_proc.so` is being built as a 64-bit binary.

Strange not hapenning on my side but I on the LLVM profile, could be that?

**Comment by MrDuartePT** [comment]
@mattst88  Now I notice I forget to add `[${MULTILIB_USEDEP}]` to NVK depedencies :sob:

**Comment by MrDuartePT** [comment]
> Oh, interesting! That seems plausible, because I don't have `bindgen` built with `ABI_X86=32`. I'll see if that fixes it.
> 
> ... Damn, no, building `bindgen` with `ABI_X86=32` didn't help.

Yes didn't help, for me also have error with gcc, it fail on `src/gallium/targets/d3dadapter9/d3dadapter9.so.1.0.0`
But It show a ton of NIR (NKV rust compiler) error, seems that something is off with gcc.
Using gcc on LLVM profile give worst error.

File: media-libs/mesa/mesa-9999.ebuild [hunk 4]
```diff
@@ -383,6 +442,11 @@ multilib_src_configure() {
 		-Db_ndebug=$(usex debug false true)
 	)
 	meson_src_configure
+
+	if ! multilib_is_native_abi && use video_cards_nvk; then
+		einfo "Applying Gentoo hack for nvk - 2/2"
```
Comment (MrDuartePT) [inline media-libs/mesa/mesa-9999.ebuild:447]: @mattst88 check if the NVK is being apply properly

Comment (MrDuartePT) [reply media-libs/mesa/mesa-9999.ebuild:447]: Well for my understanding NIR dependents on the rust subprojects, it very strange it compiling one binaries as 64bits if the submodules are 32bits

Comment (mattst88) [reply media-libs/mesa/mesa-9999.ebuild:447]: I see

```
 * Applying Gentoo hack for nvk - 1/2                                                                                                                                                                                                  
...
 * Applying Gentoo hack for nvk - 2/2                                                                                                                                                                                                  
```

And I see

```
framework /var/tmp/portage/media-libs/mesa-9999 # cat temp/rust_fix.ini 
[binaries]
rust = ['rustc', '--target=i686-unknown-linux-gnu']
```

and `--native-file /var/tmp/portage/media-libs/mesa-9999/temp/rust_fix.ini` is passed to the 32-bit configure step.

Comment (MrDuartePT) [reply media-libs/mesa/mesa-9999.ebuild:447]: Very strange NIR is compiling in 64 bits mode

**Comment by mattst88** [comment]
```
framework /var/tmp/portage/media-libs/mesa-9999 # diff -u ./work/mesa-9999-abi_x86_32.x86/build.ninja.orig ./work/mesa-9999-abi_x86_32.x86/build.ninja
--- ./work/mesa-9999-abi_x86_32.x86/build.ninja.orig    2024-05-08 12:48:23.925657928 -0400
+++ ./work/mesa-9999-abi_x86_32.x86/build.ninja 2024-05-08 12:48:23.952324089 -0400
@@ -9,7 +9,7 @@
 # Rules for compiling.
 
 rule rust_COMPILER_FOR_BUILD
- command = rustc --target=i686-unknown-linux-gnu -C linker=x86_64-pc-linux-gnu-gcc -C link-arg=-m32 -C link-arg=-mfpmath=sse $ARGS $in
+ command = rustc -C linker=x86_64-pc-linux-gnu-gcc -C link-arg=-mfpmath=sse $ARGS $in
  deps = gcc
  depfile = $targetdep
  description = Compiling Rust source $in
```

Looks odd?

**Comment by mattst88** [comment]
It sure seems like this hack, which removes `--target=i686-unknown-linux-gnu` and `-C link-arg=-m32` from the flags used for compiling rust, could certainly cause the rust code to not be compiled for 32-bit.

Are you totally sure that the multilib build works for you? After a successful build, what is the output of `file /usr/lib/libvulkan_nouveau.so`?

**Comment by MrDuartePT** [comment]
> It sure seems like this hack, which removes `--target=i686-unknown-linux-gnu` and `-C link-arg=-m32` from the flags used for compiling rust, could certainly cause the rust code to not be compiled for 32-bit.
> 
> Are you totally sure that the multilib build works for you? After a successful build, what is the output of `file /usr/lib/libvulkan_nouveau.so`?

Yes the output above look really odd. As you can see I have multilib enable (required for steam):
```bash
[ebuild   R   *] media-libs/mesa-9999::gentoo  USE="X d3d9 llvm lm-sensors opencl (opengl) osmesa proprietary-codecs vaapi vdpau vulkan wayland (zink) zstd -debug (-selinux) -test -unwind -valgrind -vulkan-overlay -xa" ABI_X86="32 (64) (-x32)" CPU_FLAGS_X86="sse2" LLVM_SLOT="(18) -15 -16 -17" VIDEO_CARDS="nouveau nvk radeonsi -d3d12 (-freedreno) -intel -lavapipe (-lima) (-panfrost) -r300 -r600 -radeon (-v3d) (-vc4) -virgl (-vivante) -vmware" 0 KiB
```

Here is the output of `/usr/lib/libvulkan_nouveau.so`:
```
mrduarte@GentooLegion ~ $ file /usr/lib/libvulkan_nouveau.so
/usr/lib/libvulkan_nouveau.so: ELF 32-bit LSB shared object, Intel 80386, version 1 (SYSV), dynamically linked, BuildID[sha1]=7b467f0a0546e1de3f14258eb8761b9d22047d48, stripped
```
As you can see is 32bits

**Comment by MrDuartePT** [comment]
> ```
> framework /var/tmp/portage/media-libs/mesa-9999 # diff -u ./work/mesa-9999-abi_x86_32.x86/build.ninja.orig ./work/mesa-9999-abi_x86_32.x86/build.ninja
> --- ./work/mesa-9999-abi_x86_32.x86/build.ninja.orig    2024-05-08 12:48:23.925657928 -0400
> +++ ./work/mesa-9999-abi_x86_32.x86/build.ninja 2024-05-08 12:48:23.952324089 -0400
> @@ -9,7 +9,7 @@
>  # Rules for compiling.
>  
>  rule rust_COMPILER_FOR_BUILD
> - command = rustc --target=i686-unknown-linux-gnu -C linker=x86_64-pc-linux-gnu-gcc -C link-arg=-m32 -C link-arg=-mfpmath=sse $ARGS $in
> + command = rustc -C linker=x86_64-pc-linux-gnu-gcc -C link-arg=-mfpmath=sse $ARGS $in
>   deps = gcc
>   depfile = $targetdep
>   description = Compiling Rust source $in
> ```
> 
> Looks odd?

It should be looking something like this:
```
rule rust_COMPILER
 command = rustc --target=i686-unknown-linux-gnu -C linker=clang-18 -C link-arg=-m32 -C link-arg=-mfpmath=sse $ARGS $in
 deps = gcc
 depfile = $targetdep
 description = Compiling Rust source $in
 ```

**Comment by MrDuartePT** [comment]
@st0nie sorry for the ping (he also use gentoo and help me mantain LLL), I don't sure if you use the default of LLVM profile, can you try compile mesa with NVK support if you don't use LLVM on your system.

@mattst88 I thinking this might be a problem related to gcc in any way, since I don't have any system with nvidia and gcc profile so I didn't test it. This is kinda a bummer since it was working last weak for you.

**Comment by mattst88** [comment]
Oh no, I updated my system and got meson-1.4.0, and because I'm testing this PR by running `ebuild ... install` I didn't realize it /o\

**Comment by MrDuartePT** [comment]
> Oh no, I updated my system and got meson-1.4.0, and because I'm testing this PR by running `ebuild ... install` I didn't realize it /o\

Thanks god was that :)

**Comment by mattst88** [comment]
Merged now! Thanks a ton. Very nice work.

And sorry for the hassle of rediscovering something you'd already sorted out /o\

**Comment by MrDuartePT** [comment]
No problem I was also forgetting about that since I have meson-1.4.0 masked (thanks for finding the menson issue)

At least this was in time of mesa 24.1 release

**Pull request closed without merging**